### PR TITLE
[TIR] Add TIR While node

### PR DESCRIPTION
**Pull request by masahi**
This is an implementation of TIR While node as discussed in RFC https://discuss.tvm.apache.org/t/rfc-add-while-loop-node-to-tir/9028. It supercedes my earlier attempt in https://github.com/apache/tvm/pull/7385.

The PR consists of
* IR node definition + boilerplate
* Minimal changes to TIR transform passes (so far only modifies `storage_rewrite.cc`, everything else uses the default visitor)
* LLVM and C source codegen
* Various test cases
* Update CUDA NMS to use while loop

Hybrid script support etc are left for future work.

Now we can write binary search succinctly as follows:
```
lo[0] = 0
hi[0] = n
v = B[i]

with ib.while_loop(lo[0] < hi[0]):
    mid = lo[0] + (hi[0] - lo[0] >> 1)
    with ib.if_scope(A[mid] < v):
        lo[0] = mid + 1
    with ib.else_scope():
        hi[0] = mid

C[i] = lo[0]
```
As another nice use of while loop, I added a test that draws a useless mandelbrot set :slightly_smiling_face:  

mandel

@tqchen @junrushao1994 @vinx13 @mbrookhart @zhiics @kevinthesun @anijain2305 @trevor-m

### CODE REVIEW

**Review by giuseros — APPROVED** [review]
LGTM, very nice addition!

**Comment by tqchen** [comment]
Thanks @masahi , before we merge it in. it would be really awesome to go through the current list of passes and check if special handling of while is needed. Some of the example passes could include(check for passes that need special IfThenElse handling)

For example, I can see the need to update following pass: 
- Vectorize (we will need to abort if the condition is vectorized)

**Comment by tqchen** [comment]
also cc @zxybazh please help to review this PR

**Review by junrushao1994 — APPROVED** [review]
Thanks for the PR! It looks good to me :-) Surprisingly it doesn't need to change any passes besides storage_rewrite :-)

**Comment by junrushao1994** [comment]
CC @spectrometerHBH: we might want to have it supported in TensorIR too, either like a syntactic sugar to opaque binding or other ways

File: include/tvm/tir/stmt_functor.h [hunk 1]
```diff
@@ -109,6 +110,7 @@ class StmtFunctor<R(const Stmt& n, Args... args)> {
     IR_STMT_FUNCTOR_DISPATCH(AttrStmtNode);
     IR_STMT_FUNCTOR_DISPATCH(IfThenElseNode);
     IR_STMT_FUNCTOR_DISPATCH(ForNode);
+    IR_STMT_FUNCTOR_DISPATCH(WhileNode);
```
Comment (tqchen) [inline include/tvm/tir/stmt_functor.h:113]: need checks through the current passes, per my comment

**Review by tqchen — CHANGES_REQUESTED** [review]

**Review by zxybazh — COMMENTED** [review]
Thanks @masahi! Looks good to me.

**Comment by masahi** [comment]
@tqchen @junrushao1994 @vinx13 

I went through passes and here is my summary:
* `VectorizeLoop`: Need to disallow a while loop inside a vectorized loop. Without it, no errors occurs during lowering but the lowered code is incorrect. Add a test case `test_vectorize_while_fail()` to make sure we error out in such cases

* `StorageAccessVisitor`: I don't understand what it does, but added a special visitor for `While` following the existing visitor for `IfThenElse`. Please check https://github.com/apache/tvm/pull/7425/commits/1e629b68b4112a01293683edc13c3e976a22a5bb

* `CoProcSync` and `LiftAttrScope`: They both have special visitor for `IfThenElse`, but I don't understand them. They are only used by VTA, for now I just error out if we find `WhileNode` there. See https://github.com/apache/tvm/pull/7425/commits/a71066d49381aae62626593c8fd76e149e1e55ed and https://github.com/apache/tvm/pull/7425/commits/00c17d921005eecc07f4300df898b9107d15ea1d

* `InjectVirtualThread`: I think we need some special handling for this, but I don't know what it should be. For now I just added a placeholder and call the base class visitor. See https://github.com/apache/tvm/pull/7425/commits/896b02fb8aba00c22696f92195d32454bd593454 and let me know what we should do here.

**Comment by masahi** [comment]
@tqchen Can you have a look?

File: src/tir/transforms/inject_virtual_thread.cc [hunk 2]
```diff
@@ -333,6 +333,12 @@ class VTInjector : public StmtExprMutator {
     }
   }
 
+  // While
+  Stmt VisitStmt_(const WhileNode* op) final {
+    // TODO(masahi): Do we need a special handling for While nodes?
```
Comment (tqchen) [inline src/tir/transforms/inject_virtual_thread.cc:338]: Let us disable for now. Likely need special handling later

**Comment by tqchen** [comment]
I left a comment for inject virtual thread, @junrushao1994 @ZihengJiang @vinx13 would be great if you can also help check the StorageAccessVisitor

**Comment by vinx13** [comment]
I've checked `StorageAccessVisitor` and it looks good to me.`InplaceOpVerifier`, `StoragePlanRewriter` also needs handling.

**Comment by masahi** [comment]
@vinx13 Ok, For `InplaceOpVerifier` I think I need to update
https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L241-L251

But I don't see how we should update `StoragePlanRewriter`. Maybe here? https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L757-L773

**Comment by tqchen** [comment]
@masahi , it would also be great for you to spend a bit more time to look into these passes :) It certainly takes more time, but  we also have more experts in TIR  passes :)

**Comment by vinx13** [comment]
@masahi For `StoragePlanRewriter`, we need to do something similar to `ForNode` https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L440-L452

**Comment by masahi** [comment]
ok, to me it's not obvious what it is doing, time for another deep dive...

**Comment by masahi** [comment]
@tqchen @vinx13 @junrushao1994 Does the behavior of `While` node wrt `StorageRewrite` below look reasonable?

In the following IR, "A" and "B" buffers, which are allocated in `For` loop, are coalesced into a one buffer, but "C" buffer, which is allocated inside `While` loop, is not:  
```
def test_parallel_alloc():
    ib = tvm.tir.ir_builder.create()
    n = te.var("n")
    with ib.for_range(0, n, name="i", kind="parallel") as i:
        with ib.for_range(0, 10, name="j") as j:
            A = ib.allocate("float32", n, name="A", scope="global")
            A[j] = A[j] + 2

        with ib.for_range(0, 10, name="j") as j:
            B = ib.allocate("float32", n, name="B", scope="global")
            B[j] = B[j] + 2

        i = ib.allocate("int32", (1,), name="i", scope="local")
        i[0] = 1
        with ib.while_loop(i[0] < 10):
            C = ib.allocate("float32", n, name="C", scope="local")
            C[i[0]] = C[i[0]] + 2
            i[0] += 1
```

```
parallel (i, 0, n) {
  // attr [A] storage_scope = "global"
  allocate A[float32 * n]
  // attr [i] storage_scope = "local"
  allocate i[int32 * 1]
  // attr [C] storage_scope = "local"
  allocate C[float32 * n]
  for (j, 0, 10) {
    A[j] = (A[j] + 2f)
  }
  for (j, 0, 10) {
    A[j] = (A[j] + 2f)
  }
  i[0] = 1
  while((i[0] < 10)){
    C[i[0]] = (C[i[0]] + 2f)
    i[0] = (i[0] + 1)
  }
}
```

In the following IR, all buffers, including the one allocated inside `While` loop, are coalesced:
```
def test_alloc_seq():
    scope_tb = "local.L0A"
    max_bits = 1024 * 1024 * 1024

    register_mem(scope_tb, max_bits)

    ib = tvm.tir.ir_builder.create()
    n = te.var("n")
    with ib.for_range(0, n, name="i") as i:
        with ib.for_range(0, 10, name="j") as j:
            A = ib.allocate("float32", 200, name="A", scope=scope_tb)
            A[j] = 1.2
        with ib.for_range(0, 10, name="j") as j:
            B = ib.allocate("float32", 200, name="B", scope=scope_tb)
            B[j] = 1.3

        i = ib.allocate("int32", (1,), name="i", scope="local")
        i[0] = 1
        with ib.while_loop(i[0] < 10):
            C = ib.allocate("float32", 200, name="C", scope=scope_tb)
            C[i[0]] = 1.4
            i[0] += 1

    body = ib.get()
```

```
// attr [A] storage_scope = "local.L0A"
allocate A[float32 * 200]
// attr [i] storage_scope = "local"
allocate i[int32 * 1]
for (i, 0, n) {
  for (j, 0, 10) {
    A[j] = 1.2f
  }
  for (j, 0, 10) {
    A[j] = 1.3f
  }
  i[0] = 1
  while((i[0] < 10)){
    A[i[0]] = 1.4f
    i[0] = (i[0] + 1)
  }
}


```

**Comment by tqchen** [comment]
@vinx13 can you please take another look at the PR and manage?

**Review by tqchen — COMMENTED** [review]
the change has addressed my previous comments. Please add testcases to transforms that touches requires special While handling to cover these passes

**Review by ZihengJiang — APPROVED** [review]
Nice work! Thanks @masahi

**Review by vinx13 — APPROVED** [review]

**Comment by masahi** [comment]
> Thanks @masahi ! the change has addressed my previous comments. Please add testcases to transforms that touches requires special While handling to cover these passes

Yes, I'm still trying to figure out what `StorageRewrite` is doing. This pass is a beast :slightly_smiling_face: I think it is doing something like storage coalescing etc, and since this is purely for optimization (I think), I'm not sure what the "failure" means in this case.

For example, we should definitely prevent invalid optimization. But I so far I'm having hard timing coming up with an example program that could fail. Anyone have any idea? Another example of failure is a missed optimization, I have an example where `For` loop coalesces but `While` loop does not https://github.com/apache/tvm/pull/7425#issuecomment-779798238

I've also added a non trivial change to `StorageAccessVisitor`, I need to look at this class and its derived classes `ThreadSyncPlanner` etc

**Comment by tqchen** [comment]
Thanks @masahi . The primary thing to address is correctness(e.g. we should not generate invalid code). It is totally fine to not optimize While well. So the test coverage is mainly to make sure that the case is handled correctly. Please let us know once you confirmed the storageaccess and rewrite, then we can revisit and merge

**Comment by mbrookhart** [comment]
@tqchen are there any specific things you still want @masahi to change? I have a large speedup for sort waiting in the wings for this PR to merge.

**Comment by masahi** [comment]
Yeah I'm still trying to figure out if `storage_rewrite` is safe for `While` loop and if not, what to do and what tests to add.

**Comment by masahi** [comment]
@tqchen @junrushao1994 @vinx13 @ZihengJiang @zxybazh 

I came to a conclusion that While node doesn't need a special handling in `storage_rewrite`.

The first observation is that even if I remove all `ForNode` handling from `StoragePlanRewriter`, all tests in `test_tir_transform_storage_rewrite.py` except [test_parallel_alloc()](https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/tests/python/unittest/test_tir_transform_storage_rewrite.py#L269) pass.

If we look at the visitor for `ForNode`, https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L440-L452
it only does something special when `attach_map_` has an entry for this node. Here comes the second observation: the only case where`attach_map_` can have an entry for `ForNode` is if this `ForNode` is a parallel for loop, due to these lines: https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L766-L772 

Together, these two handler for `ForNode` lift allocation inside an inner loop and attach merged allocation under the parallel loop scope (via `MakeAttach` function at https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L447). This is tested in `parallel_alloc`. For other kinds of `For` loop, a merged allocation is placed at the global scope, see https://github.com/apache/tvm/blob/7340c02d0efe0f5eb5692fb9f4cc7573c5d056cb/src/tir/transforms/storage_rewrite.cc#L457-L461.  

Since `While` node doesn't involve threading, I think we can always lift allocation done inside `While` loop into the global scope. That means `WhileNode` should be handled in the same way non-parallel `ForNode` are handled, i.e. we don't need a special handling logic for `WhileNode`.

I think I nailed it, thoughts?

**Comment by vinx13** [comment]
@masahi You are right, thanks for looking into this

**Comment by junrushao1994** [comment]
That makes sense to me. Thanks for diving deep into this issue!

**Comment by masahi** [comment]
cc @tqchen please take a look

**Review by tqchen — APPROVED** [review]

**Comment by tqchen** [comment]
@masahi you are right that the MakeAttach is only needed for parallel for loop, where we can nolonger lift the memory to the outside(otherwise the memory won't be thread local)

**Comment by tqchen** [comment]
@junrushao1994 @vinx13 please help to manage the PR

File: tests/python/unittest/test_tir_ir_builder.py [hunk 3]
```diff
@@ -173,9 +173,383 @@ def check_target(target):
...
+        ref = np.searchsorted(a_np, b_np)
+        tvm.testing.assert_allclose(c.asnumpy(), ref)
+
+    check_target("llvm", searchsorted_ir_cpu)
+    check_target("cuda", searchsorted_ir_gpu)
+    check_target("nvptx", searchsorted_ir_gpu)
+
+
+def test_vectorize_while_fail():
+    """A while loop inside a vectorized loop should fail."""
```
Comment (tqchen) [inline tests/python/unittest/test_tir_ir_builder.py:499]: Please move to test_tir_transform_vectorize.py

**Review by tqchen — CHANGES_REQUESTED** [review]
One minor comment

File: tests/python/unittest/test_tir_ir_builder.py (hunk 3 above)
Comment (masahi) [reply tests/python/unittest/test_tir_ir_builder.py]: done

**Review by tqchen — APPROVED** [review]

**Review by junrushao1994 — APPROVED** [review]

**Comment by masahi** [comment]
@junrushao1994 @vinx13 @tqchen ready to merge...!!

**Comment by vinx13** [comment]
Thanks everyone @masahi @tqchen @junrushao1994 @giuseros @zxybazh @ZihengJiang

**Comment by junrushao1994** [comment]
Really awesome work!!!

**Comment by masahi** [comment]
Thank you very much for the reviews!!

**Pull request merged**